include/Type.hpp

```diff
@@ -63,125 +50,239 @@ namespace cytnx {
+  // This **MUST** match the ordering of __type::__pybind_type
+  using Type_list =
+    std::tuple<void, cytnx_complex128, cytnx_complex64, cytnx_double, cytnx_float, cytnx_int64,
+               cytnx_uint64, cytnx_int32, cytnx_uint32, cytnx_int16, cytnx_uint16, cytnx_bool>;
+
+  // The number of supported types
+  constexpr int N_Type = std::tuple_size_v<Type_list>;
+  constexpr int N_fType = 5;
+
+  // The friendly name of each type
+  template <typename T>
+  constexpr char* Type_names;
+  template <>
+  constexpr const char* Type_names<void> = "Void";
+  template <>
+  constexpr const char* Type_names<cytnx_complex128> = "Complex Double (Complex Float64)";
+  template <>
+  constexpr const char* Type_names<cytnx_complex64> = "Complex Float (Complex Float32)";
+  template <>
+  constexpr const char* Type_names<cytnx_double> = "Double (Float64)";
+  template <>
+  constexpr const char* Type_names<cytnx_float> = "Float (Float32)";
+  template <>
+  constexpr const char* Type_names<cytnx_int64> = "Int64";
+  template <>
+  constexpr const char* Type_names<cytnx_uint64> = "Uint64";
+  template <>
+  constexpr const char* Type_names<cytnx_int32> = "Int32";
+  template <>
+  constexpr const char* Type_names<cytnx_uint32> = "Uint32";
+  template <>
+  constexpr const char* Type_names<cytnx_int16> = "Int16";
+  template <>
+  constexpr const char* Type_names<cytnx_uint16> = "Uint16";
+  template <>
+  constexpr const char* Type_names<cytnx_bool> = "Bool";
+
+  // The corresponding Python enumeration name
+  template <typename T>
+  constexpr char* Type_enum_name;
+  template <>
+  constexpr const char* Type_enum_name<void> = "Void";
+  template <>
+  constexpr const char* Type_enum_name<cytnx_complex128> = "ComplexDouble";
+  template <>
+  constexpr const char* Type_enum_name<cytnx_complex64> = "ComplexFloat";
+  template <>
+  constexpr const char* Type_enum_name<cytnx_double> = "Double";
+  template <>
+  constexpr const char* Type_enum_name<cytnx_float> = "Float";
+  template <>
+  constexpr const char* Type_enum_name<cytnx_int64> = "Int64";
+  template <>
+  constexpr const char* Type_enum_name<cytnx_uint64> = "Uint64";
+  template <>
+  constexpr const char* Type_enum_name<cytnx_int32> = "Int32";
+  template <>
+  constexpr const char* Type_enum_name<cytnx_uint32> = "Uint32";
+  template <>
+  constexpr const char* Type_enum_name<cytnx_int16> = "Int16";
+  template <>
+  constexpr const char* Type_enum_name<cytnx_uint16> = "Uint16";
+  template <>
+  constexpr const char* Type_enum_name<cytnx_bool> = "Bool";
 
   struct Type_struct {
-    std::string name;
-    // char name[35];
+    const char* name;  // char* is OK here, it is only ever initialized from a string literal
+    const char* enum_name;
     bool is_unsigned;
     bool is_complex;
     bool is_float;
     bool is_int;
     unsigned int typeSize;
   };
 
-  static const int N_Type = 12;
-  const int N_fType = 5;
+  template <typename T>
+  struct Type_struct_t {
```

Comment:
Is is possible unpack the `Type_list` directly in `make_type_array_helper` and not to use this helper structure?

Reply:
Yes, it could be replaced by some helper function, eg
```c++
    template <typename T>
    constexpr Type_struct make_type_struct() {
        return {
            Type_names<T>,
            Type_enum_name<T>,
            std::is_unsigned_v<T>,
            is_complex_v<T>,
            std::is_floating_point_v<T> || is_complex_floating_point_v<T>,
            std::is_integral_v<T> && !std::is_same_v<T, bool>,
            type_size<T>
        };
    }
    
    template <typename Tuple, std::size_t... Indices>
    constexpr auto make_type_array_helper(std::index_sequence<Indices...>) {
        return std::array<Type_struct, sizeof...(Indices)>{
            make_type_struct<std::tuple_element_t<Indices, Tuple>>()...};
    }
```

but I was thinking that Type_struct_t is actually an interesting class, and potentially more useful than Type_struct itself.  `is_float(unsigned int)` works for runtime calculations, but if you have some type T at compile time and you want to find that out, then a good way is `Type_struct_t<T>::is_float`. But that name isn't great, perhaps call it `Type_traits` instead?  Or simply `Traits` ? And then `Type_struct` can move into `internal` namespace, as it is an implementation detail of the `is_float(unsigned int)` etc family of functions.

I was also thinking to add `is_real`, meaning a floating point type but not complex.

Reply:
I get your intention. However, at the compile time, the type is known, so we can directly use type traits template in STL to check the type. Using STL may make the code a little bit longer, but it is straightforward to people who hasn't checked `Type.hpp`.

For example, in case where needs `Type_struct_t<T>::is_float`, I will use `std::is_floating_point_v<T> || is_complex_floating_point_v<T>` or just `std::is_floating_point_v<T> || is_complex_v<T>` instead. BTW, the name of `is_float()` is ambiguous with the meaning of `std::is_same_v<T, float>`. If someone needs to get the display name of the type or the size of type, `Type_names<T>` and `sizeof(T)` satisfy the requirement.

Reply:
What is easier to type, and less likely to go wrong, `Traits<T>::is_float`, or `std::is_floating_point_v<T> || is_complex_floating_point_v<T>` ?  And it could easily go wrong, for example your suggestion `std::is_floating_point_v<T> || is_complex_v<T>` will give the wrong answer if `T` is `std::complex<int>` (OK, that is not standard C++, and I've never used it, but I think some compilers allow it.)

The idea is that `Traits<T>::is_XXX` is guaranteed to map exactly same value as `Type.is_XXX(dtype)`, and with the same spelling.

For `is_float`, that is the name of the existing definition.  I have no objection to change it, but it is used elsewhere:
```
grep -r 'Type.is_float('
src/linalg/Lanczos_Exp.cpp:      cytnx_error_msg(!Type.is_float(Hop->dtype()),
src/linalg/Lanczos_ER.cpp:        !Type.is_float(Hop->dtype()),
src/linalg/Arnoldi.cpp:        !Type.is_float(Hop->dtype()),
src/linalg/Arnoldi.cpp:        !Type.is_float(Hop->dtype()),
src/linalg/Lanczos_Gnd.cpp:        !Type.is_float(Hop->dtype()),
src/linalg/Lanczos_Gnd_Ut.cpp:        !Type.is_float(Hop->dtype()),
bm_tests/linalg/Lanczos_bm.cpp:    if (Type.is_float(A.dtype())) {
bm_tests/linalg/Lanczos_bm.cpp:    if (Type.is_float(HEff.dtype())) {
tests/linalg_test/Arnoldi_test.cpp:    if (Type.is_float(this->dtype())) {
tests/linalg_test/Arnoldi_test.cpp:    if (Type.is_float(this->dtype())) {
tests/linalg_test/Lanczos_Exp_test.cpp:    if (Type.is_float(A.dtype())) {
tests/linalg_test/Lanczos_Exp_test.cpp:    if (Type.is_float(HEff.dtype())) {
tests/linalg_test/Arnoldi_Ut_test.cpp:    if (Type.is_float(this->dtype())) {
tests/gpu/linalg_test/Arnoldi_test.cpp:    if (Type.is_float(this->dtype())) {
tests/gpu/linalg_test/Arnoldi_test.cpp:    if (Type.is_float(this->dtype())) {
```

Reply:
> The idea is that Traits<T>::is_XXX is guaranteed to map exactly same value as Type.is_XXX(dtype), and with the same spelling.

I agree with this statement, which is why I prefer using std directly. One potential issue with `is_complex_v<T>` arises if we decide to support complex number types beyond `std::complex<float>` and `std::complex<double>`. This is because `is_complex_v` is defined by us. To clarify, the suggestion for `is_complex_v` is only valid if we limit our support to these two types of complex numbers in the foreseeable future. 

My philosophy is not to reproduce the functions that we feel it has a bad name like `is_float`. Instead, those functions should be marked as deprecated. If we really want to pack `std::is_floating_point_v<T> || is_complex_floating_point_v<T>`,

```cpp
template <class T>
using is_float = std::is_floating_point_v<T> || is_complex_floating_point_v<T>;
```

is shorter and also satisfies the requirement.

I am okay with `Type_struct_t`. I just wanted to throw an idea to avoid wrapping std functions too much. Feel free to resolve this thread.

Reply:
That is very neat!

I think `Type_struct` still has its uses though, because it collects all of the traits used by the library into a single location, rather than spreading them out into a bunch of individual traits. I can imagine it is useful, eg
```c++
void ShowTypeInfo(std::ostream& out, const Type_struct& t) {
   out << "Type:  " << t.name 
       << "\nId: " << t.cy_typeid 
       << "is_complex: " << t.is_complex
       << ....
}
```
If `type_struct` doesn't exist then you would need to implement this quite differently.

You probably still want some helper functions, since `is_int(1)` is much easier to type than `TypeTrait<cytnx::is_int>(1)`, but the helper functions could be implemented in terms of `TypeTrait`. And you definitely need `cytnx::is_int`, rather than `std::is_integral`. The reason is that the traits defined by the library are a *point of customization*, where the library, or users who are extending the library with their own types, can determine their own behavior.  `std::is_integral` is only true for the built-in C++ integral types, which, for historical reasons, includes `bool`.  Cytnx does not regard bool as integral, which is a reasonable choice for numerical algorithm purposes.  (Hence `static constexpr bool is_int = std::is_integral_v<T> && !std::is_same_v<T, bool>;`.)

if you have your own `bigint` class, for example, that you wanted to add to the library, how do you do it? You are not allowed to provide your own specializations of classes in `namespace std`, so you cannot cause `std::is_integral<bigint>` to be `std::true_type`. But you *can* specialize `cytnx::is_int<bigint>` and thereby configure the library to treat your `bigint` class as an integer.

Similarly for floating point types. I was thinking some more about this, and I think the distinctions that are useful for numerical algorithms are probably something like `is_real<T>`, `is_complex<T>` (used to select symmetric vs hermitian, or orthogonal vs unitary) and `is_real_or_complex<T>`.  There are many other traits that one could imagine, eg for rational arithmetic.  These are all points of customization where the library (and users writing library extensions) can choose how customized types should behave.  Even for the built-in types, none of these map directly onto the `std::` traits (except `is_real<T>` is probably the same as `std::is_floating_point_v<T>` the builtin types).